In [2]:
#import libraries
import numpy as np
import pandas as pd

In [3]:
arr = np.array([10, 20, 30, 40, 50, 60, 70])
print(arr)
print('\n Average number', arr.mean())

[10 20 30 40 50 60 70]

 Average number 40.0


In [4]:
#indexing
print(arr[3])
print(arr[-2])
print(arr[6])
print(arr[-4])

40
60
70
40


In [5]:
#slicing
print(arr[1:6])
print(arr[6:])
print(arr[:7])
print(arr[-6:-1])
print(arr[7:])
print(arr[-1:-6])

[20 30 40 50 60]
[70]
[10 20 30 40 50 60 70]
[20 30 40 50 60]
[]
[]


In [6]:
#2D indexing and slicing
matrix = np.array([
    [1, 2, 3],
    [4, 5, 6],
    [7, 8, 9]
])

print(matrix[0, 1])
print(matrix[1, :])
print(matrix[1:2, 2])
print(matrix[2, :2])
print(matrix[1:2, 0:2])

2
[4 5 6]
[6]
[7 8]
[[4 5]]


In [7]:
#change 1D array to 2D array
arr2 = np.arange(1, 13)
reshape = arr2.reshape(4, 3)

print('Original array', arr2)
print('Reshape array\n', reshape)

# change 2D array to 1D array
flatten = reshape.flatten()
print('Flatten array', flatten)

Original array [ 1  2  3  4  5  6  7  8  9 10 11 12]
Reshape array
 [[ 1  2  3]
 [ 4  5  6]
 [ 7  8  9]
 [10 11 12]]
Flatten array [ 1  2  3  4  5  6  7  8  9 10 11 12]


In [8]:
#working with basic mathematical operators
arr3 = np.array([1, 2, 3, 4, 5])
arr4 = np.array([10, 20, 30, 40, 50])

#basic operators
print(arr3 + arr4)
print(arr3 - arr4)
print(arr3 * arr4)
print(arr4 / arr3)

#scaler operator
print(arr4**2)
print(arr4//3)

[11 22 33 44 55]
[ -9 -18 -27 -36 -45]
[ 10  40  90 160 250]
[10. 10. 10. 10. 10.]
[ 100  400  900 1600 2500]
[ 3  6 10 13 16]


In [9]:
#working with matrix using dot(product)
matrix_1 = np.array([
    [1, 2],
    [3, 4]
])

matrix_2 = np.array([
    [5, 6],
    [7, 8]
])

dot = np.dot(matrix_1, matrix_2)
print('Dot product\n', dot)

multiply = matrix_1 * matrix_2
print('Multiply\n', multiply)

Dot product
 [[19 22]
 [43 50]]
Multiply
 [[ 5 12]
 [21 32]]


In [10]:
np.random.seed(42) #for reproducibility
days = np.arange(1, 31)
steps = np.random.randint(3456, 12034, size=30)

In [11]:
data = pd.DataFrame({
    'Days': days,
    'Steps': steps
})

data

Days  Steps
0      1  10726
1      2   4316
2      3   8846
3      4   8647
4      5   9190
5      6   9721
6      7   3922
7      8   7882
8      9   9034
9     10  11778
10    11   5141
11    12   4225
12    13  10405
13    14   5889
14    15   8767
15    16   8507
16    17   9876
17    18   4640
18    19   8011
19    20   6841
20    21   9852
21    22   6014
22    23  11305
23    24   5503
24    25   6203
25    26   3645
26    27   6190
27    28   6461
28    29   8114
29    30   5355

In [18]:
#basic stats
mean_steps = np.mean(steps)
std_steps = np.std(steps)
print('Average step:', mean_steps)
print('Median step:', np.median(steps))
print('Minimum step:', np.min(steps))
print('Maximum step:', np.max(steps))
print('Standard deviation:', std_steps)

Average step: 7500.2
Median step: 7946.5
Minimum step: 3645
Maximum step: 11778
Standard deviation: 2309.770715893679


In [13]:
#goal analysis for days exceeding 10,000 steps
goals = np.where(steps > 10000)[0] + 1
goals

array([ 1, 10, 13, 23])

In [14]:
#weekly analysis (split into 4 weeks)
weeks = np.array_split(steps, 4)
week_total = [print(week.sum()) for week in weeks]
week_avg = [print(week.mean()) for week in weeks]

63250
63746
56539
41471
7906.25
7968.25
8077.0
5924.428571428572


In [17]:
#Rolling moving average
window = 7
moving_avg = np.convolve(steps, np.ones(window)/window, mode='valid')
print('Rolling moving average\n', moving_avg[:7])

Rolling moving average
 [7909.71428571 7503.42857143 8177.42857143 8596.28571429 8095.42857143
 7386.14285714 7483.85714286]


In [23]:
#Anomaly detection
lower_bound = mean_steps - 2*std_steps
upper_bound = mean_steps + 2*std_steps

outliers = np.where((steps < lower_bound) | (steps > upper_bound))[0] + 1

print('Anomaly detection (low/high):', outliers)

Anomaly detection (low/high): []


Your average daily steps is 7,500, meaning that on a typical day you walk around seventy-five hundred steps.

The median is 7,946, a bit higher than the mean, which suggests most days are on the more active side.

Your lowest day is 3,645 steps, likely a low-movement day, while your highest day reaches 11,778, marking your most active day of the month. The standard deviation is 2,309, showing that your activity varies a lot — some days are very active, others much lighter.